unit/mctest_multib_allgather.c

```diff
@@ -961,13 +1046,209 @@ test_L_domain_allgather_dir0_cuts2_par()
+    array_local[bI] = mkarr(false, basis.num_basis, local_ranges_ext[bI]->volume);
+    gkyl_array_shiftc(array_local[bI], sqrt(pow(2,ndim)), 0); // Sets es_energy_fac=1.
+    if (num_ranks > 1) 
+      gkyl_array_scale(array_local[bI], 0.5*my_rank);
+    else
+      gkyl_array_scale_range(array_local[bI], 100.0*bid, local_ranges[bI]);
+    array_global[bI] = mkarr(false,  basis.num_basis, global_ranges_ext[bI]->volume);
+  }
+
+
+  struct gkyl_multib_comm_conn **mbcc_send = gkyl_malloc(num_local_blocks * sizeof(struct gkyl_multib_comm_conn *));
+  struct gkyl_multib_comm_conn **mbcc_recv = gkyl_malloc(num_local_blocks * sizeof(struct gkyl_multib_comm_conn *));
+  for (int bI= 0; bI<num_local_blocks; bI++) {
+      int bid = local_blocks[bI];
+      gkyl_rrobin_decomp_getranks(round_robin_decomp, bid, rank_list);
+      int brank = -1;
+      for (int i=0; i<branks[bid]; ++i)
+        if (rank_list[i] == my_rank) brank = i;
+
+      mbcc_send[bI] = gkyl_multib_comm_conn_new_send_from_connections(bid, brank, nghost, nconnected[bid], block_list[bid], dir, decomp);
+      mbcc_recv[bI] = gkyl_multib_comm_conn_new_recv_from_connections(bid, brank, nghost, nconnected[bid], block_list[bid], dir, decomp);
+
+      for (int ns=0; ns<mbcc_send[bI]->num_comm_conn; ++ns) {
+        // need to get the actual rank that owns this cut
+        int rank_idx = mbcc_send[bI]->comm_conn[ns].rank;
+        gkyl_rrobin_decomp_getranks(round_robin_decomp, mbcc_send[bI]->comm_conn[ns].block_id, rank_list);
+        mbcc_send[bI]->comm_conn[ns].rank = rank_list[rank_idx];
+        // Make range a subrange
+        mbcc_send[bI]->comm_conn[ns].range = *local_ranges[bI];
+      }
+      for (int nr=0; nr<mbcc_recv[bI]->num_comm_conn; ++nr) {
+        // need to get the actual rank that owns this cut
+        int rank_idx = mbcc_recv[bI]->comm_conn[nr].rank;
+        gkyl_rrobin_decomp_getranks(round_robin_decomp, mbcc_recv[bI]->comm_conn[nr].block_id, rank_list);
+        mbcc_recv[bI]->comm_conn[nr].rank = rank_list[rank_idx];
+        // Make range a subrange
+        gkyl_sub_range_init(&mbcc_recv[bI]->comm_conn[nr].range, global_ranges_ext[bI], mbcc_recv[bI]->comm_conn[nr].range.lower, mbcc_recv[bI]->comm_conn[nr].range.upper);
+      }
+
+  }
+
+  int stat = gkyl_multib_comm_conn_array_transfer(comm, num_local_blocks, local_blocks, mbcc_send, mbcc_recv, array_local, array_global);
+
+
+  for (int bI=0; bI<num_local_blocks; ++bI) {
+    struct gkyl_rect_grid grid;
+    int cells[2];
+    cells[0] = gkyl_range_shape(global_ranges[bI],0);
+    cells[1] = gkyl_range_shape(global_ranges[bI],1);
+    double gridlo[2] = {0.0,0.0};
+    double l0 = ((double) cells[0])/4.0;
+    double l1  = ((double) cells[1])/8.0;
+    double gridup[2] = {l0,l1 };
+    gkyl_rect_grid_init(&grid, 2, gridlo, gridup, cells);
+    char str[50];
+    sprintf(str, "lb%d_r%d.gkyl",bI, my_rank);
+    gkyl_grid_sub_array_write(&grid, global_ranges[bI], 0, array_global[bI], str);
```

Comment:
this isn't really a unit test if it requires the user to plot the result and check by eye. Can you implement an automatic correctness test?

Reply:
If it's too difficult to make it for any decomp (I suspect it is, and it may be unnecessary) just make it work for at least one decomp, and check that the test is run that way (we do that for MPI/NCCL tests).